packages/jaeger-ui/src/components/TracePage/TraceTimelineViewer/duck.js

```diff
@@ -97,6 +105,70 @@ function childrenToggle(state, { payload }) {
   return { ...state, childrenHiddenIDs };
 }
 
+function shouldDisableCollapse(allSpans, hiddenSpansIds) {
+  const allParentSpans = allSpans.filter(s => s.hasChildren);
+  return allParentSpans.length === hiddenSpansIds.size;
+}
+
+function expandAll(state) {
+  const childrenHiddenIDs = new Set();
+  return { ...state, childrenHiddenIDs };
+}
+
+function collapseAll(state, { payload }) {
+  const { spans } = payload;
+  if (shouldDisableCollapse(spans, state.childrenHiddenIDs)) {
+    return state;
+  }
+  const childrenHiddenIDs = spans.reduce((res, s) => {
```

Comment:
Just curious, why prefer `reduce` over `filter`?

```
const childrenHiddenIDs = new Set(spans.filter(sp => sp.hasChildren));
```
Applies generally to these reducers.